Release notes 8.12 (#12396)
coq/coq: doc/sphinx/changes.rst

   .. include:: ../unreleased.rst

Version 8.12
------------

Summary of changes
~~~~~~~~~~~~~~~~~~

Changes in 8.12+beta1
~~~~~~~~~~~~~~~~~~~~~

> **Specification language, type inference**

Reviewer: Given that there are 1000 lines describing the changes, how about making them easy to scan/jump to by here adding a list of hyperlinks to the subsections that follow OR making them (sub-)* sections so they appear in the TOC?

Author: Making them subsections doesn't make them appear in the TOC in the left-bar (too many levels) and I tried both and I prefer the appearance that the current state results in.

Author: Note that there will be a summary written by @mattam82 with some direct links to a selection of changelog entries.

Reviewer: > Making them subsections doesn't make them appear in the TOC in the left-bar (too many levels) and I tried both and I prefer the appearance that the current state results in.

We should figure out, though not in this PR, a better way to handle this because the issue arises in other long chapters or sections.  I expect the TOC should be configurable somehow.  Manually adding hyperlinks at the beginning of the section is easy enough, but isn't self-maintaining.  There ought to be a directive to make it self maintaining.  WDYT?

Author: This is actually supported in Sphinx and does look great for this changelog, so I will push a new version including this.

Reviewer: Looks good!

No way to thumbs-up on some comments? :-(

  `#11797 <https://github.com/coq/coq/pull/11797>`_,
  `#11913 <https://github.com/coq/coq/pull/11913>`_,
  `#11958 <https://github.com/coq/coq/pull/11958>`_
  `#11960 <https://github.com/coq/coq/pull/11960>`_
  and `#11961 <https://github.com/coq/coq/pull/11961>`_, by Jim
  Fehrle, with reviews of Théo Zimmermann).

**Infrastructure and dependencies**

- **Changed:**
  Minimal versions of dependencies for building the reference manual:
  now requires Sphinx 2.3.1+, sphinx_rtd_theme 0.4.3+ and
  sphinxcontrib-bibtex 0.4.2+
  (`#12224 <https://github.com/coq/coq/pull/12224>`_,
  by Jim Fehrle and Théo Zimmermann).
- **Removed:**
  Python 2 is not longer required in any part of the codebase
  (`#11245 <https://github.com/coq/coq/pull/11245>`_,
  by Emilio Jesus Gallego Arias).

**Miscellaneous**

- **Added:**
  Support for better extraction of strings in OCaml and Haskell:
  `ExtOcamlNativeString` provides bindings from the Coq `String` type to
  the OCaml `string` type, and string literals can be extracted to literals,
  both in OCaml and Haskell (`#10486
  <https://github.com/coq/coq/pull/10486>`_, by Xavier Leroy, with help from
  Maxime Dénès, review by Hugo Herbelin).

> - **Added:**

Reviewer: Maybe better in the "Tactics" section.

Author: Agreed.

  the OCaml `string` type, and string literals can be extracted to literals,
  both in OCaml and Haskell (`#10486
  <https://github.com/coq/coq/pull/10486>`_, by Xavier Leroy, with help from
  Maxime Dénès, review by Hugo Herbelin).
- **Fixed:**
  In Haskell extraction with ``ExtrHaskellString``, equality comparisons on
  strings and characters are now guaranteed to be uniquely well-typed, even in
  very polymorphic contexts under ``unsafeCoerce``; this is achieved by adding
  type annotations to the extracted code, and by making ``ExtrHaskellString``
  export ``ExtrHaskellBasic`` (`#12263
  <https://github.com/coq/coq/pull/12263>`_, by Jason Gross, fixes `#12257
  <https://github.com/coq/coq/issues/12257>`_ and `#12258
  <https://github.com/coq/coq/issues/12258>`_).

Reference manual
^^^^^^^^^^^^^^^^

- **Changed:**
  The reference manual has been restructured to get a more logical
  organization.  In the new version, there are fewer top-level
  chapters, and, in the HTML format, chapters are split into smaller
  pages.  This is still a work in progress and further restructuring
  is expected in the next versions of Coq

>   (`CEP#43 <https://github.com/coq/ceps/pull/43>`_, implemented in

Reviewer: Just a thought: it might friendlier to link to `https://github.com/coq/ceps/pull/43/files` which shows the text of the CEP.  Or, better, the formatted document if that's accessible.  The latter would be good for the 8.12 announcement email, which can include a CEP link for further information.

Author: The top post of the CEP contains the link to the rendered version already.

  (`#9803 <https://github.com/coq/coq/pull/9803>`_,
  by Laurent Théry and Michael Soegtrop).

**Reference manual**

- **Changed:**

>   The reference manual has been restructured.  In the new version,
>   there are less top-level chapters, and, in the HTML format, chapters
>   are split on multiple pages.  This is still a work-in-progress and
>   further restructuring is expected in the next versions of Coq

Reviewer: I'd include a little bit of why as well as what was done.

Mention that sections were reordered to give a more logical organization.

For HTML, split on multiple pages to improve usability.  Or say "split into smaller pages".

Wording corrections:
```suggestion
  The reference manual has been restructured.  In the new version,
  there are fewer top-level chapters, and, in the HTML format, chapters
  are split into multiple pages.  This is still a work in progress and
  further restructuring is expected in the next versions of Coq
```

Author: Motivations were manyfolds (as expressed in the CEP, which this changelog entry links to) so it seems hard to summarize it shortly, but I'll try.

  (`CEP#43 <https://github.com/coq/ceps/pull/43>`_, implemented in
  `#11601 <https://github.com/coq/coq/pull/11601>`_,
  `#11871 <https://github.com/coq/coq/pull/11871>`_,
  `#11914 <https://github.com/coq/coq/pull/11914>`_,
  `#12148 <https://github.com/coq/coq/pull/12148>`_,
  `#12172 <https://github.com/coq/coq/pull/12172>`_,
  `#12239 <https://github.com/coq/coq/pull/12239>`_
  and `#12330 <https://github.com/coq/coq/pull/12330>`_,
  effort inspired by Matthieu Sozeau, led by Théo Zimmermann, with
  help and reviews of Jim Fehrle, Clément Pit-Claudel and others).
- **Changed:**

>   Most of the grammar is now presented using the same box mechanism as
>   was already used to present commands and tactics since Coq 8.8 and
>   which is documented in :ref:`syntax-conventions`

Reviewer: ```suggestion
  Most of the grammar is now presented using the notation mechanism that
  has been used to present commands and tactics since Coq 8.8,
  which is documented in :ref:`syntax-conventions`
```

  (`#11183 <https://github.com/coq/coq/pull/11183>`_,
  `#11314 <https://github.com/coq/coq/pull/11314>`_,
  `#11423 <https://github.com/coq/coq/pull/11423>`_,
  `#11705 <https://github.com/coq/coq/pull/11705>`_,
  `#11718 <https://github.com/coq/coq/pull/11718>`_,
  `#11720 <https://github.com/coq/coq/pull/11720>`_
  and `#11961 <https://github.com/coq/coq/pull/11961>`_, by Jim

>   Fehrle, with reviews of Théo Zimmermann).

Reviewer: This is idiomatic.  Applies in several other places.

```suggestion
  Fehrle, reviewed by Théo Zimmermann).
```

- **Added:**
  A glossary of terms and an index of attributes
  (`#11869 <https://github.com/coq/coq/pull/11869>`_,
  `#12150 <https://github.com/coq/coq/pull/12150>`_
  and `#12224 <https://github.com/coq/coq/pull/12224>`_,
  by Jim Fehrle and Théo Zimmermann,
  with reviews of Clément Pit-Claudel)
- **Added:**

>   A selector that allows to switch between versions of the reference

Reviewer: ```suggestion
  A selector that allows switching between versions of the reference
```

  manual (`#12286 <https://github.com/coq/coq/pull/12286>`_, by
  Clément Pit-Claudel).
- **Fixed:**

>   A large part of the documented syntax has been fixed to correspond
>   more accurately to what is actually implemented, thanks to an
>   internal `doc_grammar` tool introduced for this purpose

Reviewer: ```suggestion
  Most of the documented syntax has been thoroughly updated to make it accurate and
  easily understood.  This was done using a semi-automated `doc_grammar`
  tool introduced for this purpose and through significant revisions to the text.
```

Still very understated--getting an accurate, readable grammar is a big step forward.
The pre-change grammar was really bad and incomplete.
Still a good bit of syntax that's not explained (e.g. in the terms chapter).

Will we mention the doc improvements (and the plans for subsequent improvements) more prominently, such as in the email announcing the release?

Author: Sorry, I missed most of your comments last time:

>Will we mention the doc improvements (and the plans for subsequent improvements) more prominently, such as in the email announcing the release?

There's a summary to be written above, which highlights a selection of changes. The documentation changes in 8.12 are clearly one of the highlights to be done.

The e-mail announcing the release will repeat these highlights. But in addition, the 8.12+beta1 announcement should include a call for beta readers of the manual and for help in improving the presentation further. I'll get back to you for help writing this call.
